itests/fevm_test.go

```diff
@@ -1436,3 +1436,107 @@ func TestEthGetBlockByNumber(t *testing.T) {
 	require.NotNil(t, pendingBlock)
 	require.True(t, pendingBlock.Number >= latest)
 }
+
+func TestEthGetTransactionByBlockHashAndIndexAndNumber(t *testing.T) {
+	ctx, cancel, client := kit.SetupFEVMTest(t)
+	defer cancel()
+
+	ethKey, ethAddr, ethFilAddr := client.EVM().NewAccount()
+	filKey, err := client.WalletNew(ctx, types.KTBLS)
+	require.NoError(t, err)
+	secpKey, err := client.WalletNew(ctx, types.KTSecp256k1)
+	require.NoError(t, err)
+
+	kit.SendFunds(ctx, t, client, ethFilAddr, types.FromFil(10))
+	kit.SendFunds(ctx, t, client, filKey, types.FromFil(10))
+	kit.SendFunds(ctx, t, client, secpKey, types.FromFil(10))
+
+	var txHashes []ethtypes.EthHash
+	var receipts []*api.EthTxReceipt
+	numTx := 3
+
+	for i := 0; i < numTx; i++ {
+		ethTx := deployContractWithEth(ctx, t, client, ethAddr, "./contracts/MultipleEvents.hex")
+		ethTx.Nonce = i
+		client.EVM().SignTransaction(ethTx, ethKey.PrivateKey)
+		ethHash := client.EVM().SubmitTransaction(ctx, ethTx)
+
+		receipt, err := client.EVM().WaitTransaction(ctx, ethHash)
+		require.NoError(t, err, "ETH transaction failed to mine")
+		require.NotNil(t, receipt)
+		require.Equal(t, ethtypes.EthUint64(1), receipt.Status)
+
+		txHashes = append(txHashes, ethHash)
+		receipts = append(receipts, receipt)
+	}
+
+	require.NotEmpty(t, receipts, "No ETH transactions were mined")
+	var blockHashes []ethtypes.EthHash
+	var blockNumbers []ethtypes.EthUint64
+	for _, receipt := range receipts {
+		blockHashes = append(blockHashes, receipt.BlockHash)
+		blockNumbers = append(blockNumbers, receipt.BlockNumber)
+	}
+
+	for _, hash := range txHashes {
+		var ethTx *ethtypes.EthTx
+		var err error
+		for _, blockHash := range blockHashes {
+			ethTx, err = client.EthGetTransactionByBlockHashAndIndex(ctx, blockHash, ethtypes.EthUint64(0))
```

Comment:
@virajbhartiya this still isn't quite doing what we need, your fundamental problem is back when you deploy the contracts.

Have a close look at TestMultipleEvents in eth_filter_test.go, in this loop: https://github.com/filecoin-project/lotus/blob/02a8b972db82eee0ff109366f57122caf7790128/itests/eth_filter_test.go#L694-L741

Note particularly the `allInSameTipset` check at the end and the fact that the loop will continue until it gets that correct. It'll keep on submitting 3 transactions (from `paramsArray`) until it can see that they all landed in the same tipset.

Your test as it is now is submitting _and_ waiting for each transaction, so they're guaranteed to land on separate tipsets so you only ever get to test index=0. Instead, you should:

1. Set up a `for {}` loop
2. Run through your `numTx` loop, only doing `SubmitTransaction` for each of them
3. Wait for each of the hashes you submitted in another loop, so we know they all landed
4. Check all of the receipts that their `BlockHash` is the same.
5. If they are all the same, break, if they are different, continue the loop.

Then you should have `numTx` transactions (messages) in a single block (tipset).

The second problem you have is that you can't guarantee that they all landed in the order you expect, _or_ that they are the only ones to land in the tipset! There is a miner in the background that you're not interacting with that's keeping the chain going and it could have got something in that tipset as well. So what I think you should do is use the `receipts[index].TransactionIndex` when you're calling `EthGetTransactionByBlockHashAndIndex` rather than a fixed number. But note also your "Index out of range" has this problem too - you don't know how many should be in that tipset, there may be more than you submitted. I think for that one you should just request a high, improbable index, like `100`.

Reply:
We need multiple transaction in the same `tipset`, since the function is designed to iterate through transactions in a single tipset - testing with just one transaction per tipset wouldn't verify that the indexing mechanism works correctly when there are multiple transactions to index through. Multiple transactions allow testing different index positions (0, 1, 2, etc.)
This verifies the function can correctly retrieve transactions at different positions within a block. so looping it in `for {}` and checking `allInSameTipset` makes sure that all land in the same `tipset`.